File: sdk/go/krasis/ai.go
package krasis

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

// AIModule handles AI question operations.
type AIModule struct {
	client *Client
}

// NewAIModule creates a new AI module.
func NewAIModule(c *Client) *AIModule {
	return &AIModule{client: c}
}

// Ask sends a question and returns the answer.
func (a *AIModule) Ask(req AskRequest) (*AskResponse, error) {
	var resp AskResponse
	err := a.client.Post("/ai/ask", req, &resp)
	return &resp, err
}

// AskStream sends a question and returns a stream of tokens.
func (a *AIModule) AskStream(ctx context.Context, req AskRequest) (<-chan string, <-chan error, error) {
	req.Stream = true

	data, err := json.Marshal(req)
	if err != nil {
		return nil, nil, err
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, a.client.baseURL+"/ai/ask/stream", bytes.NewReader(data))
	if err != nil {
		return nil, nil, err
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if token := a.client.Token(); token != "" {
		httpReq.Header.Set("Authorization", "Bearer "+token)
	}

	resp, err := a.client.http.Do(httpReq)
	if err != nil {
		return nil, nil, err
	}

	if resp.StatusCode != http.StatusOK {
		defer resp.Body.Close()
		return nil, nil, &APIError{StatusCode: resp.StatusCode, Message: fmt.Sprintf("HTTP %d", resp.StatusCode)}
	}

	tokenCh := make(chan string)
	errCh := make(chan error, 1)

	go func() {
		defer resp.Body.Close()
		defer close(tokenCh)
		defer close(errCh)

		scanner := bufio.NewScanner(resp.Body)
		var lastEvent string
		for scanner.Scan() {
			line := scanner.Text()
			if strings.HasPrefix(line, "event: ") {
				lastEvent = strings.TrimPrefix(line, "event: ")
				continue
			}
			if strings.HasPrefix(line, "data: ") {
				payload := strings.TrimPrefix(line, "data: ")
				if lastEvent == "token" {
					var d struct {
						Token string `json:"token"`
					}
					if err := json.Unmarshal([]byte(payload), &d); err == nil && d.Token != "" {
						select {
						case tokenCh <- d.Token:
						case <-ctx.Done():
							return
						}
					}
				} else if lastEvent == "done" {
					return
				}
			}
		}
	}()

	return tokenCh, errCh, nil
}

// ListConversations returns all conversations.
func (a *AIModule) ListConversations() ([]Conversation, error) {
	var convs []Conversation
	err := a.client.Get("/ai/conversations", &convs)
	return convs, err
}

// GetMessages returns all messages in a conversation.
func (a *AIModule) GetMessages(conversationID string) ([]Message, error) {
	var msgs []Message
	err := a.client.Get("/ai/conversations/"+conversationID+"/messages", &msgs)
	return msgs, err
}

// CreateConversation creates a new conversation.
func (a *AIModule) CreateConversation(title, model string) (*Conversation, error) {
	body := map[string]string{}
	if title != "" {
		body["title"] = title
	}
	if model != "" {
		body["model"] = model
	}
	var conv Conversation
	err := a.client.Post("/ai/conversations", body, &conv)
	return &conv, err
}

File: sdk/python/krasis/ai.py
"""AI module with streaming support."""

from __future__ import annotations

import json
from collections.abc import AsyncIterator
from typing import Any

import aiohttp

from krasis.client import Client
from krasis.types import AskRequest, AskResponse, Conversation, Message


class AIModule:
    """AI question and conversation operations."""

    def __init__(self, client: Client) -> None:
        self._client = client

    async def ask(self, req: AskRequest) -> AskResponse:
        """Send a question and return the answer."""
        body = self._request_to_dict(req)
        data = await self._client.post("/ai/ask", body=body)
        return AskResponse(**data)

    async def ask_stream(self, req: AskRequest) -> AsyncIterator[str]:
        """Send a question and yield tokens as they arrive via SSE."""
        req.stream = True
        body = self._request_to_dict(req)

        import json as _json

        session = await self._client._get_session()
        url = f"{self._client.base_url}/ai/ask/stream"
        headers: dict[str, str] = {"Content-Type": "application/json"}
        if self._client.token:
            headers["Authorization"] = f"Bearer {self._client.token}"

        async with session.post(url, headers=headers, data=_json.dumps(body)) as resp:
            if resp.status != 200:
                body_text = await resp.text()
                raise Exception(f"HTTP {resp.status}: {body_text}")

            last_event = ""
            async for line in resp.content:
                line_str = line.decode("utf-8").strip()
                if line_str.startswith("event: "):
                    last_event = line_str[7:]
                    continue
                if line_str.startswith("data: "):
                    payload = line_str[6:]
                    if last_event == "token":
                        try:
                            d = json.loads(payload)
                            token = d.get("token", "")
                            if token:
                                yield token
                        except json.JSONDecodeError:
                            continue
                    elif last_event == "done":
                        return

    async def list_conversations(self) -> list[Conversation]:
        """Return all conversations."""
        data = await self._client.get("/ai/conversations")
        return [Conversation(**c) for c in data]

    async def get_messages(self, conversation_id: str) -> list[Message]:
        """Return all messages in a conversation."""
        data = await self._client.get(f"/ai/conversations/{conversation_id}/messages")
        return [Message(**m) for m in data]

    async def create_conversation(self, title: str = "", model: str = "") -> Conversation:
        """Create a new conversation."""
        body: dict[str, str] = {}
        if title:
            body["title"] = title
        if model:
            body["model"] = model
        data = await self._client.post("/ai/conversations", body=body)
        return Conversation(**data)

    @staticmethod
    def _request_to_dict(req: AskRequest) -> dict[str, Any]:
        result: dict[str, Any] = {"question": req.question}
        if req.conversation_id:
            result["conversation_id"] = req.conversation_id
        if req.model:
            result["model"] = req.model
        result["stream"] = req.stream
        return result
